# Heart Disease Prediction

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import sklearn
from sklearn.preprocessing import MinMaxScaler
%matplotlib inline

In [2]:
heart = pd.read_csv('/Users/nikhilkuniyil/Desktop/HeartDisease_Project/heart.csv')

In [4]:
heart.head(10)

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   52    1   0       125   212    0        1      168      0      1.0      2   
1   53    1   0       140   203    1        0      155      1      3.1      0   
2   70    1   0       145   174    0        1      125      1      2.6      0   
3   61    1   0       148   203    0        1      161      0      0.0      2   
4   62    0   0       138   294    1        1      106      0      1.9      1   
5   58    0   0       100   248    0        0      122      0      1.0      1   
6   58    1   0       114   318    0        2      140      0      4.4      0   
7   55    1   0       160   289    0        0      145      1      0.8      1   
8   46    1   0       120   249    0        0      144      0      0.8      2   
9   54    1   0       122   286    0        0      116      1      3.2      1   

   ca  thal  target  
0   2     3       0  
1   0     3       0  
2   0     3       0  
3   1     3       0  
4   3     2       0  
5   0     2       1  
6   3     1       0  
7   1     3       0  
8   0     3       0  
9   2     2       0

In [5]:
heart.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1025 entries, 0 to 1024
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       1025 non-null   int64  
 1   sex       1025 non-null   int64  
 2   cp        1025 non-null   int64  
 3   trestbps  1025 non-null   int64  
 4   chol      1025 non-null   int64  
 5   fbs       1025 non-null   int64  
 6   restecg   1025 non-null   int64  
 7   thalach   1025 non-null   int64  
 8   exang     1025 non-null   int64  
 9   oldpeak   1025 non-null   float64
 10  slope     1025 non-null   int64  
 11  ca        1025 non-null   int64  
 12  thal      1025 non-null   int64  
 13  target    1025 non-null   int64  
dtypes: float64(1), int64(13)
memory usage: 112.2 KB


In [6]:
heart.describe()

age          sex           cp     trestbps        chol  \
count  1025.000000  1025.000000  1025.000000  1025.000000  1025.00000   
mean     54.434146     0.695610     0.942439   131.611707   246.00000   
std       9.072290     0.460373     1.029641    17.516718    51.59251   
min      29.000000     0.000000     0.000000    94.000000   126.00000   
25%      48.000000     0.000000     0.000000   120.000000   211.00000   
50%      56.000000     1.000000     1.000000   130.000000   240.00000   
75%      61.000000     1.000000     2.000000   140.000000   275.00000   
max      77.000000     1.000000     3.000000   200.000000   564.00000   

               fbs      restecg      thalach        exang      oldpeak  \
count  1025.000000  1025.000000  1025.000000  1025.000000  1025.000000   
mean      0.149268     0.529756   149.114146     0.336585     1.071512   
std       0.356527     0.527878    23.005724     0.472772     1.175053   
min       0.000000     0.000000    71.000000     0.000000     0.000000   
25%       0.000000     0.000000   132.000000     0.000000     0.000000   
50%       0.000000     1.000000   152.000000     0.000000     0.800000   
75%       0.000000     1.000000   166.000000     1.000000     1.800000   
max       1.000000     2.000000   202.000000     1.000000     6.200000   

             slope           ca         thal       target  
count  1025.000000  1025.000000  1025.000000  1025.000000  
mean      1.385366     0.754146     2.323902     0.513171  
std       0.617755     1.030798     0.620660     0.500070  
min       0.000000     0.000000     0.000000     0.000000  
25%       1.000000     0.000000     2.000000     0.000000  
50%       1.000000     0.000000     2.000000     1.000000  
75%       2.000000     1.000000     3.000000     1.000000  
max       2.000000     4.000000     3.000000     1.000000

In [15]:
heart.shape

(1025, 14)

In [16]:
heart['target'].value_counts()

1    526
0    499
Name: target, dtype: int64

A value of 1 means that the patient has heart disease, while a value of 0 means that the patient does not have heart disease.

In this dataset, 526 patients have heart disease while 499 do not.

In [18]:
heart.groupby(heart['target']).mean()

age       sex        cp    trestbps        chol       fbs  \
target                                                                    
0       56.569138  0.827655  0.482966  134.106212  251.292585  0.164329   
1       52.408745  0.570342  1.378327  129.245247  240.979087  0.134981   

         restecg     thalach     exang   oldpeak     slope        ca      thal  
target                                                                          
0       0.456914  139.130261  0.549098  1.600200  1.166333  1.158317  2.539078  
1       0.598859  158.585551  0.134981  0.569962  1.593156  0.370722  2.119772

## Logistic Regression

In [3]:
X = heart.drop(columns='target')
y = heart['target']

In [4]:
scaler = MinMaxScaler()
X = scaler.fit_transform(X)

### Train-Test Split

In [5]:
from sklearn.model_selection import train_test_split

In [6]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=101)

### Train Logistic Regression Model

In [7]:
from sklearn.linear_model import LogisticRegression

In [8]:
logr = LogisticRegression()
logr.fit(X_train,y_train)

LogisticRegression()

### Model Evaluation

#### Accuracy Score

In [9]:
from sklearn.metrics import classification_report, accuracy_score, confusion_matrix

In [10]:
#accuracy score on training data
training_predictions = logr.predict(X_train)
accuracy = accuracy_score(training_predictions,y_train)
print(f"Accuracy of training data: {accuracy}")

Accuracy of training data: 0.8437935843793585


In [11]:
#accuracy score on testing data
testing_predictions = logr.predict(X_test)
accuracy = accuracy_score(testing_predictions,y_test)
print(f"Accuracy of testing data: {accuracy}")

Accuracy of testing data: 0.8246753246753247


#### Classification Report

In [12]:
print(classification_report(testing_predictions,y_test))

              precision    recall  f1-score   support

           0       0.74      0.88      0.80       124
           1       0.91      0.79      0.84       184

    accuracy                           0.82       308
   macro avg       0.82      0.83      0.82       308
weighted avg       0.84      0.82      0.83       308



In [13]:
print(confusion_matrix(testing_predictions,y_test))

[[109  15]
 [ 39 145]]


### Heart Disease Predictive System

In [15]:
heart_data = np.array([58,0,0,100,248,0,0,122,0,1.0,1,0,2])

# re-shape since we only are predicting for the instance above
input_reshape = heart_data.reshape(1,-1)

scaled_data = scaler.fit_transform(input_reshape)
prediction = logr.predict(scaled_data)

if prediction[0] == 1:
    print("Patient has heart disease")
else:
    print("Patient does not have heart disease")

Patient has heart disease


Data points were sampled from the 6th row of the 'heart' dataset. Our logistic regression model has therefore prediced
correctly that a patient with the following attributes equivalent to that of the 6th row of the 'heart_data' dataset,
does in fact have heart disease.

## Saving Model

In [71]:
import pickle

In [72]:
pickle.dump(logr, open('model.pkl','wb'))

model = pickle.load(open('model.pkl','rb'))